storage/src/main/java/tech/pegasys/teku/storage/store/Store.java

```diff
@@ -212,17 +233,12 @@ public SignedBeaconBlock getSignedBlock(Bytes32 blockRoot) {
 
   @Override
   public Optional<SignedBlockAndState> getBlockAndState(final Bytes32 blockRoot) {
-    readLock.lock();
-    try {
-      final SignedBeaconBlock block = blocks.get(blockRoot);
-      final BeaconState state = block_states.get(blockRoot);
-      if (block == null || state == null) {
-        return Optional.empty();
-      }
-      return Optional.of(new SignedBlockAndState(block, state));
-    } finally {
-      readLock.unlock();
+    final SignedBeaconBlock block = getSignedBlock(blockRoot);
+    final Optional<BeaconState> state = getOrGenerateBlockState(blockRoot);
+    if (block == null || state.isEmpty()) {
+      return Optional.empty();
     }
+    return Optional.of(new SignedBlockAndState(block, state.get()));
```

Comment:
Would it be worth bailing out before attempting to recreate the state if `block == null`?  Performance-wise it won't make any significant difference but it feels weird to ask for something we already know we can't generate.  I think you could then just use a `map`:
```suggestion
    final SignedBeaconBlock block = getSignedBlock(blockRoot);
    if (block == null) {
      return Optional.empty();
    }
    return getOrGenerateBlockState(blockRoot).map(state -> new SignedBlockAndState(block, state));
```